Add retryEAGAIN callback function

mmarchini opened the pull request:
Add a synchronous callback for when we encounter EAGAIN errors. The
function must return `true` or `false` to inform sonic-boom if it should
continue retrying or if it should bubble up the error it encounters.

Fixes: https://github.com/pinojs/sonic-boom/issues/65

---

The implementation here is different than the ones discussed in the issue. Adding a maxEAGAIN/maxFlushRetries prove challeging because of `SonicBoom.release` (hard to keep track of attempts with `SonicBoom.release` and a mix of `fs.write` and `fs.writeSync` in the code). This alternative approach gives control to the user to decide if it should retry or not. With this, flushSync will throw if retryEAGAIN returns false. I'm open to discuss other approaches but wanted to have something as a starting point first.

mcollina reviewed (commented):
The approach is ok for me.

I think we should pass the length of the current buffer as input. A safe implementation for this function could return false if the current buffer is over a threshold.

We would need to make sure to surface this to pino.

mmarchini commented:
> I think we should pass the length of the current buffer as input. A safe implementation for this function could return false if the current buffer is over a threshold.

I like this idea, let me update the PR.

> We would need to make sure to surface this to pino.

`pino.destination` passes all options to sonic-boom, so I think we're covered there.

mmarchini commented:
Updated the pull request to also pass the length of the current buffer (or more precisely, the current buffer sonic-boom is trying to write) as input. Is that what you meant, or should I pass `this._buf.length` instead on both calls?

mcollina commented:
I meant `._buf`, but both would be useful.

mmarchini commented:
Updated the pull request to pass the length of the buffer `sonic-boom` tried to write as well as the length of the buffer with data that sonic-boom didn't try to write (which only happens on async when it exceeds MAX_WRITE, but it can happen and is quite relevant to the scenarios described in #65).

I think this addresses all comments so far.

mcollina reviewed (approved):
lgtm

jsumners commented on README.md:
```diff
@@ -63,10 +63,16 @@ The options are:
+    true sonic-boom will retry the operation, otherwise it will bubble the
+    error. `err` is the error that caused this function to be called,
+    `writeBufferLen` is the length of the buffer sonic-boom tried to write, and
+    `remainingBufferLen` is the lenght of the remaining buffer sonic-boom didn't try to write.
```
```suggestion
    `remainingBufferLen` is the length of the remaining buffer sonic-boom didn't try to write.
```

mmarchini commented:
I rebased it before fetching, so I was on an outdated sonic-boom version 🤦🏻‍♀️. I'll fix it and push again soon.

mmarchini commented:
the last commit should account for the atomic async changes and should fix the tests

mcollina commented:
This PR and the atomic async changes are a great combo. I can't wait to have them in production.
